File: ProGO/backend/ml/python/intent_classifier.py
"""
Intent Classification using Scikit-learn
Advanced ML model for chatbot intent recognition
"""

import json
import pickle
import numpy as np
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.preprocessing import LabelEncoder
import joblib
from typing import Dict, List, Tuple
import os

class IntentClassifier:
    """Advanced Intent Classification with Random Forest"""
    
    def __init__(self, model_path='models/intent_classifier.pkl',
                 vectorizer_path='models/intent_vectorizer.pkl'):
        self.model_path = model_path
        self.vectorizer_path = vectorizer_path
        self.model = None
        self.vectorizer = None
        self.label_encoder = LabelEncoder()
        self.intent_labels = []
        
    def get_training_data(self) -> Tuple[List[str], List[str]]:
        """Get comprehensive training data"""
        
        training_data = {
            'greeting': [
                'hello', 'hi', 'hey', 'good morning', 'good afternoon',
                'good evening', 'greetings', 'hi there', 'hello there',
                'hey there', 'whats up', 'how are you', 'howdy'
            ],
            'attendance': [
                'what is my attendance', 'show my attendance', 'check attendance',
                'my attendance percentage', 'how much attendance do i have',
                'attendance status', 'am i present today', 'how many days present',
                'attendance record', 'view attendance', 'total attendance',
                'attendance report', 'show attendance details', 'check my presence',
                'how many classes attended', 'attendance percentage', 'my present days'
            ],
            'marks': [
                'what are my marks', 'show my marks', 'check marks', 'my scores',
                'what did i score', 'exam results', 'my grades', 'show grades',
                'marks obtained', 'view marks', 'result', 'my performance in exams',
                'how much did i score', 'marks in subject', 'subject marks',
                'exam marks', 'test scores', 'my marks card'
            ],
            'fees': [
                'what are my fees', 'fee status', 'how much fees', 'pending fees',
                'fees due', 'payment status', 'fee payment', 'outstanding fees',
                'total fees', 'fees amount', 'check fees', 'fee details',
                'how much do i owe', 'fee structure', 'tuition fees', 'fees paid'
            ],
            'schedule': [
                'what is my schedule', 'show timetable', 'class schedule', 'todays classes',
                'class timings', 'when is next class', 'timetable', 'class timing',
                'what classes today', 'schedule for tomorrow', 'weekly schedule',
                'class routine', 'subject schedule', 'show my timetable'
            ],
            'performance': [
                'how am i performing', 'my performance', 'academic performance',
                'am i doing well', 'performance analysis', 'how is my progress',
                'cgpa details', 'my cgpa', 'gpa status', 'overall performance',
                'semester performance', 'academic progress', 'performance report',
                'how am i doing academically', 'my grades overview'
            ],
            'backlog': [
                'do i have backlogs', 'backlog status', 'how many backlogs',
                'failed subjects', 'backlog details', 'kt details', 'arrears',
                'which subjects failed', 'backlog list', 'failed exams',
                'subjects to clear', 'pending subjects'
            ],
            'exam': [
                'when is exam', 'exam schedule', 'exam dates', 'upcoming exams',
                'exam timetable', 'next exam', 'exam details', 'test schedule',
                'when is next exam', 'exam calendar', 'examination dates',
                'exam information', 'test dates'
            ],
            'subject': [
                'what subjects am i studying', 'my subjects', 'subject list',
                'subjects this semester', 'which subjects', 'subject details',
                'subjects enrolled', 'course subjects', 'subject information',
                'syllabus', 'course details', 'subjects in current semester'
            ],
            'help': [
                'help', 'what can you do', 'help me', 'how to use this',
                'what are your features', 'assist me', 'guide me',
                'how does this work', 'need help', 'support', 'what can i ask'
            ],
            'thank_you': [
                'thank you', 'thanks', 'appreciate it', 'thanks a lot',
                'thank you very much', 'grateful', 'thanks so much',
                'that helps', 'perfect thanks', 'great thank you'
            ],
            'general_query': [
                'tell me about', 'what is', 'explain', 'how does', 'why',
                'can you tell me', 'i want to know', 'information about',
                'details about', 'what do you know about'
            ]
        }
        
        # Expand training data
        texts = []
        labels = []
        
        for intent, phrases in training_data.items():
            for phrase in phrases:
                texts.append(phrase)
                labels.append(intent)
                # Add variations
                texts.append(phrase + '?')
                labels.append(intent)
                texts.append(phrase.title())
                labels.append(intent)
        
        return texts, labels
    
    def train(self, texts: List[str] = None, labels: List[str] = None):
        """
        Train the intent classifier
        
        Args:
            texts: List of training texts (optional, uses default if None)
            labels: List of intent labels (optional, uses default if None)
        """
        # Get training data
        if texts is None or labels is None:
            texts, labels = self.get_training_data()
        
        print(f"Training with {len(texts)} samples across {len(set(labels))} intents")
        
        # Encode labels
        y = self.label_encoder.fit_transform(labels)
        self.intent_labels = list(self.label_encoder.classes_)
        
        # Create TF-IDF features
        self.vectorizer = TfidfVectorizer(
            max_features=1000,
            ngram_range=(1, 3),  # Unigrams, bigrams, trigrams
            min_df=1,
            lowercase=True,
            strip_accents='unicode'
        )
        
        X = self.vectorizer.fit_transform(texts)
        
        # Split data
        X_train, X_test, y_train, y_test = train_test_split(
            X, y, test_size=0.2, random_state=42, stratify=y
        )
        
        # Train Random Forest
        self.model = RandomForestClassifier(
            n_estimators=100,
            max_depth=50,
            min_samples_split=2,
            random_state=42,
            n_jobs=-1
        )
        
        print("Training Random Forest classifier...")
        self.model.fit(X_train, y_train)
        
        # Evaluate
        train_score = self.model.score(X_train, y_train)
        test_score = self.model.score(X_test, y_test)
        
        print(f"\nTraining Accuracy: {train_score:.4f}")
        print(f"Testing Accuracy: {test_score:.4f}")
        
        # Cross-validation
        cv_scores = cross_val_score(self.model, X, y, cv=5)
        print(f"Cross-Validation Accuracy: {cv_scores.mean():.4f} (+/- {cv_scores.std() * 2:.4f})")
        
        # Detailed evaluation
        y_pred = self.model.predict(X_test)
        print("\nClassification Report:")
        print(classification_report(y_test, y_pred, 
                                   target_names=self.intent_labels,
                                   zero_division=0))
        
        # Save models
        self.save()
        
        return {
            'train_accuracy': float(train_score),
            'test_accuracy': float(test_score),
            'cv_accuracy': float(cv_scores.mean()),
            'cv_std': float(cv_scores.std()),
            'n_samples': len(texts),
            'n_intents': len(self.intent_labels)
        }
    
    def predict(self, text: str, return_probabilities: bool = False) -> Dict:
        """
        Predict intent for given text
        
        Args:
            text: Input text
            return_probabilities: Whether to return all probabilities
            
        Returns:
            Dictionary with intent, confidence, and optionally all probabilities
        """
        if self.model is None or self.vectorizer is None:
            self.load()
        
        # Transform text
        X = self.vectorizer.transform([text])
        
        # Predict
        prediction = self.model.predict(X)[0]
        probabilities = self.model.predict_proba(X)[0]
        
        # Get intent and confidence
        intent = self.label_encoder.inverse_transform([prediction])[0]
        confidence = float(probabilities[prediction])
        
        result = {
            'intent': intent,
            'confidence': confidence
        }
        
        # Add alternative intents
        if len(probabilities) > 1:
            # Get top 3 alternatives
            top_indices = np.argsort(probabilities)[-3:][::-1]
            alternatives = [
                {
                    'intent': self.label_encoder.inverse_transform([idx])[0],
                    'confidence': float(probabilities[idx])
                }
                for idx in top_indices if idx != prediction and probabilities[idx] > 0.1
            ]
            result['alternatives'] = alternatives
        
        if return_probabilities:
            result['all_probabilities'] = {
                self.label_encoder.inverse_transform([i])[0]: float(prob)
                for i, prob in enumerate(probabilities)
            }
        
        return result
    
    def save(self):
        """Save model and vectorizer to disk"""
        os.makedirs(os.path.dirname(self.model_path), exist_ok=True)
        
        # Save model
        joblib.dump(self.model, self.model_path)
        print(f"Model saved to {self.model_path}")
        
        # Save vectorizer and label encoder
        with open(self.vectorizer_path, 'wb') as f:
            pickle.dump({
                'vectorizer': self.vectorizer,
                'label_encoder': self.label_encoder,
                'intent_labels': self.intent_labels
            }, f)
        print(f"Vectorizer saved to {self.vectorizer_path}")
    
    def load(self):
        """Load model and vectorizer from disk"""
        if not os.path.exists(self.model_path):
            raise FileNotFoundError(f"Model not found at {self.model_path}. Please train first.")
        
        # Load model
        self.model = joblib.load(self.model_path)
        
        # Load vectorizer and label encoder
        with open(self.vectorizer_path, 'rb') as f:
            data = pickle.load(f)
            self.vectorizer = data['vectorizer']
            self.label_encoder = data['label_encoder']
            self.intent_labels = data['intent_labels']
        
        print(f"Model loaded from {self.model_path}")


# CLI Interface
if __name__ == '__main__':
    import sys
    
    classifier = IntentClassifier()
    
    if len(sys.argv) > 1:
        command = sys.argv[1]
        
        if command == 'train':
            # Train the model
            stats = classifier.train()
            print(json.dumps(stats, indent=2))
            
        elif command == 'predict':
            # Predict intent for text
            if len(sys.argv) < 3:
                print("Usage: python intent_classifier.py predict <text>")
                sys.exit(1)
            
            text = ' '.join(sys.argv[2:])
            result = classifier.predict(text, return_probabilities=True)
            print(json.dumps(result, indent=2))
            
        elif command == 'test':
            # Test the model
            test_cases = [
                "What is my attendance?",
                "Show me my marks",
                "How much fees do I have to pay?",
                "Hello",
                "Thank you",
                "When is the next exam?",
                "Do I have any backlogs?",
                "How am I performing this semester?"
            ]
            
            results = []
            for text in test_cases:
                result = classifier.predict(text)
                results.append({
                    'text': text,
                    'intent': result['intent'],
                    'confidence': result['confidence']
                })
                print(f"Text: {text}")
                print(f"  Intent: {result['intent']} ({result['confidence']:.3f})")
                print()
            
            print(json.dumps(results, indent=2))
    else:
        print("Usage:")
        print("  python intent_classifier.py train")
        print("  python intent_classifier.py predict <text>")
        print("  python intent_classifier.py test")

File: ProGO/backend/ml/intentClassifier.js
const brain = require('brain.js');
const fs = require('fs');
const path = require('path');
const natural = require('natural');

class IntentClassifier {
  constructor() {
    // Use NeuralNetwork instead of LSTM to avoid GPU requirements
    this.network = new brain.NeuralNetwork({
      hiddenLayers: [128, 64],
      learningRate: 0.01,
      errorThresh: 0.005
    });
    this.modelPath = path.join(__dirname, 'models', 'intent-model.json');
    this.conversationHistoryPath = path.join(__dirname, 'models', 'conversation-history.json');
    this.isTrained = false;
    this.tokenizer = new natural.WordTokenizer();
    this.stemmer = natural.PorterStemmer;
    
    // Context tracking for better intent recognition
    this.contextHistory = [];
    this.maxContextLength = 5;
    
    // Confidence threshold
    this.confidenceThreshold = 0.6;
  }

  // Enhanced training data for intent classification
  getTrainingData() {
    return [
      // Overall Attendance Intents (expanded)
      { input: 'show overall attendance', output: 'overall_attendance' },
      { input: 'what is my child attendance', output: 'overall_attendance' },
      { input: 'total attendance percentage', output: 'overall_attendance' },
      { input: 'attendance stats', output: 'overall_attendance' },
      { input: 'how much attendance', output: 'overall_attendance' },
      { input: 'check attendance', output: 'overall_attendance' },
      { input: 'attendance report', output: 'overall_attendance' },
      { input: 'overall attendance record', output: 'overall_attendance' },
      { input: 'total attendance', output: 'overall_attendance' },
      { input: 'attendance summary', output: 'overall_attendance' },
      { input: 'how many classes attended', output: 'overall_attendance' },
      { input: 'attendance percentage', output: 'overall_attendance' },
      { input: 'show me the attendance', output: 'overall_attendance' },
      { input: 'what is attendance', output: 'overall_attendance' },
      { input: 'can i see attendance', output: 'overall_attendance' },

      // Subject-wise Attendance Intents (expanded)
      { input: 'subject wise attendance', output: 'subject_attendance' },
      { input: 'attendance for each subject', output: 'subject_attendance' },
      { input: 'show subject attendance', output: 'subject_attendance' },
      { input: 'attendance by subject', output: 'subject_attendance' },
      { input: 'individual subject attendance', output: 'subject_attendance' },
      { input: 'mathematics attendance', output: 'subject_attendance' },
      { input: 'subject attendance details', output: 'subject_attendance' },
      { input: 'attendance in all subjects', output: 'subject_attendance' },
      { input: 'per subject attendance', output: 'subject_attendance' },
      { input: 'subject based attendance', output: 'subject_attendance' },
      { input: 'physics attendance', output: 'subject_attendance' },
      { input: 'chemistry attendance', output: 'subject_attendance' },
      { input: 'computer science attendance', output: 'subject_attendance' },
      { input: 'attendance details for each class', output: 'subject_attendance' },

      // Academic Status Intents (expanded)
      { input: 'academic status', output: 'academic_status' },
      { input: 'check backlogs', output: 'academic_status' },
      { input: 'any backlogs', output: 'academic_status' },
      { input: 'failed subjects', output: 'academic_status' },
      { input: 'academic standing', output: 'academic_status' },
      { input: 'show status', output: 'academic_status' },
      { input: 'incomplete subjects', output: 'academic_status' },
      { input: 'how many backlogs', output: 'academic_status' },
      { input: 'pending subjects', output: 'academic_status' },
      { input: 'cleared subjects', output: 'academic_status' },
      { input: 'status report', output: 'academic_status' },
      { input: 'academic record', output: 'academic_status' },
      { input: 'do they have any failures', output: 'academic_status' },
      { input: 'subjects to clear', output: 'academic_status' },

      // Academic Performance Intents (expanded)
      { input: 'show performance', output: 'academic_performance' },
      { input: 'what is cgpa', output: 'academic_performance' },
      { input: 'check grades', output: 'academic_performance' },
      { input: 'academic performance', output: 'academic_performance' },
      { input: 'semester marks', output: 'academic_performance' },
      { input: 'show results', output: 'academic_performance' },
      { input: 'grade report', output: 'academic_performance' },
      { input: 'current cgpa', output: 'academic_performance' },
      { input: 'gpa details', output: 'academic_performance' },
      { input: 'marks obtained', output: 'academic_performance' },
      { input: 'exam results', output: 'academic_performance' },
      { input: 'how are the marks', output: 'academic_performance' },
      { input: 'performance report', output: 'academic_performance' },
      { input: 'show me grades', output: 'academic_performance' },
      { input: 'what are the scores', output: 'academic_performance' },
      { input: 'semester wise performance', output: 'academic_performance' },

      // Upcoming Exams Intents (expanded)
      { input: 'upcoming exams', output: 'upcoming_exams' },
      { input: 'exam schedule', output: 'upcoming_exams' },
      { input: 'when is next exam', output: 'upcoming_exams' },
      { input: 'exam dates', output: 'upcoming_exams' },
      { input: 'test schedule', output: 'upcoming_exams' },
      { input: 'show exams', output: 'upcoming_exams' },
      { input: 'examination timetable', output: 'upcoming_exams' },
      { input: 'exam calendar', output: 'upcoming_exams' },
      { input: 'future exams', output: 'upcoming_exams' },
      { input: 'next examination', output: 'upcoming_exams' },
      { input: 'exam information', output: 'upcoming_exams' },
      { input: 'test dates', output: 'upcoming_exams' },
      { input: 'when are exams', output: 'upcoming_exams' },
      { input: 'assessment schedule', output: 'upcoming_exams' },

      // Fee Payment Intents (expanded)
      { input: 'fee status', output: 'fee_status' },
      { input: 'payment status', output: 'fee_status' },
      { input: 'check fees', output: 'fee_status' },
      { input: 'fee payment', output: 'fee_status' },
      { input: 'pending fees', output: 'fee_status' },
      { input: 'tuition fee status', output: 'fee_status' },
      { input: 'fee details', output: 'fee_status' },
      { input: 'fees paid', output: 'fee_status' },
      { input: 'outstanding fees', output: 'fee_status' },
      { input: 'payment pending', output: 'fee_status' },
      { input: 'fee due', output: 'fee_status' },
      { input: 'payment information', output: 'fee_status' },
      { input: 'how much fees pending', output: 'fee_status' },
      { input: 'unpaid fees', output: 'fee_status' },

      // Notifications Intents (expanded)
      { input: 'show notifications', output: 'notifications' },
      { input: 'announcements', output: 'notifications' },
      { input: 'latest updates', output: 'notifications' },
      { input: 'any new announcements', output: 'notifications' },
      { input: 'college notifications', output: 'notifications' },
      { input: 'recent news', output: 'notifications' },
      { input: 'important updates', output: 'notifications' },
      { input: 'new notifications', output: 'notifications' },
      { input: 'college updates', output: 'notifications' },
      { input: 'what is new', output: 'notifications' },
      { input: 'any announcements', output: 'notifications' },
      { input: 'school news', output: 'notifications' },
      { input: 'important notices', output: 'notifications' },

      // Faculty Contact Intents (expanded)
      { input: 'faculty contact', output: 'faculty_contact' },
      { input: 'teacher contact details', output: 'faculty_contact' },
      { input: 'how to contact faculty', output: 'faculty_contact' },
      { input: 'class advisor details', output: 'faculty_contact' },
      { input: 'teacher information', output: 'faculty_contact' },
      { input: 'faculty email', output: 'faculty_contact' },
      { input: 'professor contact', output: 'faculty_contact' },
      { input: 'contact teacher', output: 'faculty_contact' },
      { input: 'faculty phone number', output: 'faculty_contact' },
      { input: 'teacher email', output: 'faculty_contact' },
      { input: 'advisor contact', output: 'faculty_contact' },
      { input: 'how to reach faculty', output: 'faculty_contact' },

      // Performance Insights Intents (expanded)
      { input: 'performance insights', output: 'performance_insights' },
      { input: 'analysis report', output: 'performance_insights' },
      { input: 'weak subjects', output: 'performance_insights' },
      { input: 'strong subjects', output: 'performance_insights' },
      { input: 'performance analysis', output: 'performance_insights' },
      { input: 'improvement areas', output: 'performance_insights' },
      { input: 'performance summary', output: 'performance_insights' },
      { input: 'strengths and weaknesses', output: 'performance_insights' },
      { input: 'where to improve', output: 'performance_insights' },
      { input: 'subject analysis', output: 'performance_insights' },
      { input: 'performance prediction', output: 'performance_insights' },
      { input: 'future performance', output: 'performance_insights' },
      { input: 'academic insights', output: 'performance_insights' },

      // Greeting Intents (expanded)
      { input: 'hello', output: 'greeting' },
      { input: 'hi', output: 'greeting' },
      { input: 'hey', output: 'greeting' },
      { input: 'good morning', output: 'greeting' },
      { input: 'good afternoon', output: 'greeting' },
      { input: 'good evening', output: 'greeting' },
      { input: 'namaste', output: 'greeting' },
      { input: 'greetings', output: 'greeting' },

      // Help Intents
      { input: 'help', output: 'help' },
      { input: 'what can you do', output: 'help' },
      { input: 'show menu', output: 'help' },
      { input: 'options', output: 'help' },
      { input: 'what are my options', output: 'help' },
      { input: 'show options', output: 'help' },
      { input: 'menu', output: 'help' },
      { input: 'commands', output: 'help' },

      // Logout Intents (expanded)
      { input: 'logout', output: 'logout' },
      { input: 'exit', output: 'logout' },
      { input: 'quit', output: 'logout' },
      { input: 'sign out', output: 'logout' },
      { input: 'end session', output: 'logout' },
      { input: 'bye', output: 'logout' },
      { input: 'goodbye', output: 'logout' },

      { input: 'sign out', output: 'logout' },
      { input: 'end session', output: 'logout' },
      { input: 'bye', output: 'logout' },
      { input: 'goodbye', output: 'logout' },
      { input: 'close', output: 'logout' },
      { input: 'done', output: 'logout' },
    ];
  }

  // Preprocess input text
  preprocessInput(input) {
    // Convert to lowercase and tokenize
    const tokens = this.tokenizer.tokenize(input.toLowerCase());
    
    // Stem tokens
    const stemmedTokens = tokens.map(token => this.stemmer.stem(token));
    
    // Join back to string
    return stemmedTokens.join(' ');
  }

  // Train the neural network with enhanced training data
  async train() {
    try {
      console.log('🧠 Training Intent Classifier...');
      const trainingData = this.getTrainingData();

      // Load existing conversation history and incorporate it
      const additionalData = this.loadConversationHistory();
      const allTrainingData = [...trainingData, ...additionalData];

      console.log(`  📊 Total training samples: ${allTrainingData.length}`);
      console.log(`  📚 Base samples: ${trainingData.length}`);
      console.log(`  💬 Conversation samples: ${additionalData.length}`);

      const startTime = Date.now();

      this.network.train(allTrainingData, {
        iterations: 3000,
        errorThresh: 0.003,
        log: true,
        logPeriod: 200,
        learningRate: 0.01,
        callback: (stats) => {
          if (stats.iterations % 500 === 0) {
            console.log(`  ⏳ Iteration ${stats.iterations}: error = ${stats.error.toFixed(6)}`);
          }
        }
      });

      const endTime = Date.now();
      const duration = ((endTime - startTime) / 1000).toFixed(2);

      this.isTrained = true;
      console.log(`✓ Intent Classifier trained successfully in ${duration}s`);

      // Save the trained model
      this.saveModel();
      return true;
    } catch (error) {
      console.error('✗ Error training Intent Classifier:', error);
      throw error;
    }
  }

  // Classify user intent with context awareness
  classify(input, context = null) {
    try {
      if (!this.isTrained) {
        this.loadModel();
      }

      // Preprocess input
      const processedInput = this.preprocessInput(input);

      // Run classification
      const intent = this.network.run(processedInput);
      
      // Update context history
      this.updateContext(input, intent);

      return intent;
    } catch (error) {
      console.error('Error classifying intent:', error);
      return 'unknown';
    }
  }

  // Get confidence score for classification with multiple possibilities
  classifyWithConfidence(input, context = null) {
    try {
      if (!this.isTrained) {
        this.loadModel();
      }

      // Preprocess input
      const processedInput = this.preprocessInput(input);

      // Run classification
      const result = this.network.run(processedInput);
      
      // Calculate confidence based on pattern matching
      const confidence = this.calculateConfidence(input, result);

      // Update context tracking for conversation flow
      this.updateContext(input, result);

      // Check if we need clarification
      const needsClarification = confidence < this.confidenceThreshold;

      return {
        intent: result,
        confidence: confidence,
        input: input,
        needsClarification: needsClarification,
        alternativeIntents: this.getAlternativeIntents(input, result)
      };
    } catch (error) {
      console.error('Error classifying with confidence:', error);
      return {
        intent: 'unknown',
        confidence: 0,
        input: input,
        needsClarification: true,
        alternativeIntents: []
      };
    }
  }

  // Calculate confidence score based on keyword matching
  calculateConfidence(input, predictedIntent) {
    const trainingData = this.getTrainingData();
    
    // Find similar examples in training data
    const similarExamples = trainingData.filter(d => d.output === predictedIntent);
    
    if (similarExamples.length === 0) {
      return 0.5; // Default confidence
    }

    // Calculate similarity with training examples
    const inputTokens = new Set(this.tokenizer.tokenize(input.toLowerCase()));
    
    let maxSimilarity = 0;
    similarExamples.forEach(example => {
      const exampleTokens = new Set(this.tokenizer.tokenize(example.input.toLowerCase()));
      const intersection = new Set([...inputTokens].filter(x => exampleTokens.has(x)));
      const union = new Set([...inputTokens, ...exampleTokens]);
      
      const similarity = intersection.size / union.size;
      maxSimilarity = Math.max(maxSimilarity, similarity);
    });

    // Convert similarity to confidence (0.5 to 1.0 range)
    return Math.min(0.5 + maxSimilarity, 1.0);
  }

  // Get alternative possible intents
  getAlternativeIntents(input, primaryIntent) {
    const alternatives = [];
    const trainingData = this.getTrainingData();
    const inputTokens = new Set(this.tokenizer.tokenize(input.toLowerCase()));

    // Group by intent
    const intentGroups = {};
    trainingData.forEach(d => {
      if (!intentGroups[d.output]) {
        intentGroups[d.output] = [];
      }
      intentGroups[d.output].push(d.input);
    });

    // Calculate similarity for each intent
    Object.keys(intentGroups).forEach(intent => {
      if (intent === primaryIntent) return;

      let maxSim = 0;
      intentGroups[intent].forEach(example => {
        const exampleTokens = new Set(this.tokenizer.tokenize(example.toLowerCase()));
        const intersection = new Set([...inputTokens].filter(x => exampleTokens.has(x)));
        const union = new Set([...inputTokens, ...exampleTokens]);
        const similarity = intersection.size / union.size;
        maxSim = Math.max(maxSim, similarity);
      });

      if (maxSim > 0.3) {
        alternatives.push({ intent, confidence: maxSim });
      }
    });

    // Sort by confidence and return top 2
    return alternatives
      .sort((a, b) => b.confidence - a.confidence)
      .slice(0, 2);
  }

  // Update context history for conversation flow
  updateContext(input, intent) {
    this.contextHistory.push({
      input,
      intent,
      timestamp: Date.now()
    });

    // Keep only recent context
    if (this.contextHistory.length > this.maxContextLength) {
      this.contextHistory.shift();
    }
  }

  // Get recent context
  getContext() {
    return this.contextHistory;
  }

  // Clear context
  clearContext() {
    this.contextHistory = [];
  }

  // Save conversation for future training
  saveConversationForTraining(input, intent, wasCorrect = true) {
    try {
      let history = [];
      
      if (fs.existsSync(this.conversationHistoryPath)) {
        const data = fs.readFileSync(this.conversationHistoryPath, 'utf8');
        history = JSON.parse(data);
      }

      history.push({
        input: input.toLowerCase(),
        output: intent,
        wasCorrect,
        timestamp: new Date().toISOString()
      });

      // Keep only last 1000 conversations
      if (history.length > 1000) {
        history = history.slice(-1000);
      }

      const modelDir = path.join(__dirname, 'models');
      if (!fs.existsSync(modelDir)) {
        fs.mkdirSync(modelDir, { recursive: true });
      }

      fs.writeFileSync(this.conversationHistoryPath, JSON.stringify(history, null, 2));
    } catch (error) {
      console.error('Error saving conversation history:', error);
    }
  }

  // Load conversation history for training
  loadConversationHistory() {
    try {
      if (fs.existsSync(this.conversationHistoryPath)) {
        const data = fs.readFileSync(this.conversationHistoryPath, 'utf8');
        const history = JSON.parse(data);
        
        // Only use correct classifications for training
        return history
          .filter(h => h.wasCorrect !== false)
          .map(h => ({ input: h.input, output: h.output }));
      }
    } catch (error) {
      console.error('Error loading conversation history:', error);
    }
    return [];
  }

  // Retrain model with new data
  async retrain() {
    console.log('🔄 Retraining model with new conversation data...');
    return await this.train();
  }

  // Save trained model to file
  saveModel() {
    try {
      const modelDir = path.join(__dirname, 'models');
      if (!fs.existsSync(modelDir)) {
        fs.mkdirSync(modelDir, { recursive: true });
      }

      const model = this.network.toJSON();
      fs.writeFileSync(this.modelPath, JSON.stringify(model));
      console.log('✓ Model saved to:', this.modelPath);
    } catch (error) {
      console.error('✗ Error saving model:', error);
    }
  }

  // Load trained model from file
  loadModel() {
    try {
      if (fs.existsSync(this.modelPath)) {
        const model = JSON.parse(fs.readFileSync(this.modelPath, 'utf8'));
        this.network.fromJSON(model);
        this.isTrained = true;
        console.log('✓ Intent Classifier model loaded');
        return true;
      } else {
        console.log('⚠ No trained model found. Please train the model first.');
        return false;
      }
    } catch (error) {
      console.error('✗ Error loading model:', error);
      return false;
    }
  }

  // Get model statistics
  getModelStats() {
    const trainingData = this.getTrainingData();
    const conversationData = this.loadConversationHistory();
    
    const intentCounts = {};
    trainingData.forEach(d => {
      intentCounts[d.output] = (intentCounts[d.output] || 0) + 1;
    });

    return {
      totalTrainingSamples: trainingData.length + conversationData.length,
      baseTrainingSamples: trainingData.length,
      conversationSamples: conversationData.length,
      uniqueIntents: Object.keys(intentCounts).length,
      isTrained: this.isTrained,
      intentDistribution: intentCounts
    };
  }
}

module.exports = new IntentClassifier();
